`jemalloc` build fails on aarch64-darwin

### Describe the bug
`python3Packages.tensorflow` does not build aarch64-darwin as of eac07edbd20ed4908b98790ba299250b5527ecdf.

### Steps To Reproduce
Steps to reproduce the behavior:
1. Check out eac07edbd20ed4908b98790ba299250b5527ecdf.
2. `nix-build -A python3Packages.tensorflow` on aarch64-darwin

### Expected behavior
The build to succeed or the package to be marked as broken.

### Screenshots
n/a

### Additional context
Full build log: https://gist.github.com/samuela/42e5da41bb7a2e32842bd3de3777fced

### Notify maintainers
<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->
@jyp @abbradar @ndl @veprbl @danieldk @timokau 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"aarch64-darwin"`
 - host os: `Darwin 21.2.0, macOS 12.1`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.4`
 - channels(samuelainsworth): `"darwin, home-manager, nixpkgs-21.11pre330734.5cb226a06c4"`
 - channels(root): `"nixpkgs-21.11pre330734.5cb226a06c4"`
 - nixpkgs: `/Users/samuelainsworth/.nix-defexpr/channels/nixpkgs`
```

Ah didn't realize this was a jemalloc failure at first. Sure, here's the result on https://github.com/NixOS/nixpkgs/commit/6bda212f691a94f5b16385a432e86b9e0c98cbf5:
```
❯ nix-build -A jemalloc
error: Package ‘compiler-rt-libc-10.0.1’ in /Users/samuelainsworth/dev/nixpkgs/pkgs/development/compilers/llvm/10/compiler-rt/default.nix:95 is marked as broken, refusing to evaluate.
```
Overriding with `NIXPKGS_ALLOW_BROKEN=1` gives me a more substantial failure:
```
...
[ 87%] Building C object lib/builtins/CMakeFiles/clang_rt.builtins_arm64_osx.dir/clear_cache.c.o
[ 87%] Linking C static library libclang_rt.builtins_arm64_osx.a
[ 87%] Built target clang_rt.builtins_arm64_osx
make: *** [Makefile:136: all] Error 2
error: builder for '/nix/store/7fxzrv6n5fglzh4wjvcg7xqd7zd43rf1-compiler-rt-libc-10.0.1.drv' failed with exit code 2;
```
Full build log: https://gist.github.com/samuela/6679addeea17f965c00b6bce9392ad7c

--- a/pkgs/development/libraries/jemalloc/default.nix
+++ b/pkgs/development/libraries/jemalloc/default.nix
@@ -1,6 +1,7 @@
 { lib
 , stdenv
 , fetchurl
+, fetchpatch
 # By default, jemalloc puts a je_ prefix onto all its symbols on OSX, which
 # then stops downstream builds (mariadb in particular) from detecting it. This
 # option should remove the prefix and give us a working jemalloc.
@@ -18,6 +19,14 @@ stdenv.mkDerivation rec {
     url = "https://github.com/jemalloc/jemalloc/releases/download/${version}/${pname}-${version}.tar.bz2";
     sha256 = "sha256-LbgtHnEZ3z5xt2QCGbbf6EeJvAU3mDw7esT3GJrs/qo=";
   };
+
+  patches = [
+    # fix tests under --with-jemalloc-prefix=, see https://github.com/jemalloc/jemalloc/pull/2340
+    (fetchpatch {
+      url = "https://github.com/jemalloc/jemalloc/commit/d00ecee6a8dfa90afcb1bbc0858985c17bef6559.patch";
+      hash = "sha256-N5i4IxGJ4SSAgFiq5oGRnrNeegdk2flw9Sh2mP0yl4c=";
+    })
+  ];
 
   # see the comment on stripPrefix
   configureFlags = []

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -19720,11 +19720,7 @@ with pkgs;
 
   jbigkit = callPackage ../development/libraries/jbigkit { };
 
-  jemalloc = callPackage ../development/libraries/jemalloc {
-    # tests fail with LLVM 11+
-    # https://github.com/jemalloc/jemalloc/issues/2091
-    stdenv = if stdenv.cc.isClang then llvmPackages_10.stdenv else stdenv;
-  };
+  jemalloc = callPackage ../development/libraries/jemalloc { };
 
   jose = callPackage ../development/libraries/jose { };
 

